src/pure/Queue.mo

```diff
@@ -129,336 +189,805 @@ module {
-      if (text.size() > 10) text #= ", ";
-      text #= f(item)
+    var first = true;
+    for (t in values queue) {
+      if (first) first := false else text #= ", ";
+      text #= f(t)
     };
-    List.forEach(queue.0, add);
-    List.forEach(queue.2, add);
     text # "]"
   };
 
-  /// Compare two queues using lexicographic ordering specified by argument function `compareItem`.
+  /// Reverse the order of elements in a queue.
+  /// This operation is cheap, it does NOT require copying the elements.
   ///
   /// Example:
-  /// ```motoko include=initialize
-  /// import Nat "mo:base/Nat";
-  ///
-  /// Queue.compare<Nat>(
-  ///   (?(1, ?(2, null)), 2, null),
-  ///   (null, 2, ?(2, ?(1, null))),
-  ///   Nat.compare
-  /// ) // => #equal
+  /// ```motoko
+  /// import Queue "mo:new-base/Queue";
+  ///
+  /// let queue = Queue.fromIter([1, 2, 3].vals());
+  /// Queue.toText(Queue.reverse(queue), Nat.toText) // => "PureQueue[3, 2, 1]"
   /// ```
   ///
-  /// Runtime: O(size(l1))
+  /// Runtime: `O(1)`
+  ///
+  /// Space: `O(1)`
+  public func reverse<T>(queue : Queue<T>) : Queue<T> = switch queue {
+    case (#empty) queue;
+    case (#one(_)) queue;
+    case (#two(x, y)) #two(y, x);
+    case (#three(x, y, z)) #three(z, y, x);
+    case (#idles(l, r)) #idles(r, l);
+    case (#rebal(#left, big, small)) #rebal(#right, big, small);
+    case (#rebal(#right, big, small)) #rebal(#left, big, small)
+  };
+
+  type Stacks<T> = (left : List<T>, right : List<T>);
+
+  module Stacks {
+    public func push<T>((left, right) : Stacks<T>, t : T) : Stacks<T> = (?(t, left), right);
+
+    public func pop<T>(stacks : Stacks<T>) : Stacks<T> = switch stacks {
+      case (?(_, leftTail), right) (leftTail, right);
+      case (null, ?(_, rightTail)) (null, rightTail);
+      case (null, null) stacks
+    };
+
+    public func first<T>((left, right) : Stacks<T>) : ?T = switch (left) {
+      case (?(h, _)) ?h;
+      case (null) do ? { right!.0 }
+    };
+
+    public func unsafeFirst<T>((left, right) : Stacks<T>) : T = switch (left) {
+      case (?(h, _)) h;
+      case (null) Option.unwrap(right).0
+    };
+
+    public func isEmpty<T>((left, right) : Stacks<T>) : Bool = List.isEmpty(left) and List.isEmpty(right);
+
+    public func size<T>((left, right) : Stacks<T>) : Nat = List.size(left) + List.size(right);
+
+    public func smallqueue<T>((left, right) : Stacks<T>) : Queue<T> = switch (left, right) {
+      case (null, null) #empty;
+      case (null, ?(x, null)) #one(x);
+      case (?(x, null), null) #one(x);
+      case (null, ?(x, ?(y, null))) #two(y, x);
+      case (?(x, null), ?(y, null)) #two(y, x);
+      case (?(x, ?(y, null)), null) #two(y, x);
+      case (null, ?(x, ?(y, ?(z, null)))) #three(z, y, x);
+      case (?(x, ?(y, ?(z, null))), null) #three(z, y, x);
+      case (?(x, ?(y, null)), ?(z, null)) #three(z, y, x);
+      case (?(x, null), ?(y, ?(z, null))) #three(z, y, x);
+      case _ (trap "Illegal smallqueue invocation")
```

Comment:
Please make this  (and any other trap) consistent with style of other traps in new-motoko-base (I think they add the name of the lib and operation that trapped)